scripts/portal-fetcher/openshift-docs-downloader.py

```diff
@@ -0,0 +1,1384 @@
+#!/usr/bin/env python3
+"""OpenShift Documentation Downloader.
+
+Downloads the HTML-single version of Red Hat OpenShift documentation for a specified version,
+preserving directory structure and maintaining a database of downloaded files.
+Supports incremental updates and advanced verification.
+"""
+
+import argparse
+import asyncio
+import datetime
+import json
+import logging
+import sqlite3
+import time
+from pathlib import Path
+from typing import Optional, Union
+from urllib.parse import urljoin, urlparse
+
+import aiohttp
+from bs4 import BeautifulSoup
+
+# Configure logging
+logging.basicConfig(
+    level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
+)
+logger = logging.getLogger(__name__)
+
+# Pure functions for URL handling
+
+
+def normalize_url(url: str) -> str:
+    """Normalize a URL by removing fragment identifiers and ensuring consistency.
+
+    Args:
+        url (str): URL to normalize
+
+    Returns:
+        str: Normalized URL
+    """
+    # Parse the URL
+    parsed = urlparse(url)
+
+    # Remove fragment identifier
+    parsed = parsed._replace(fragment="")
+
+    # Build the path without trailing index if present
+    path = parsed.path
+    path = path.removesuffix("/index")
+    if not path.endswith("/"):
+        path = path + "/"
+
+    # Reassemble the URL
+    normalized = parsed._replace(path=path).geturl()
+
+    return normalized
+
+
+def is_likely_external_link(url: str) -> bool:
+    """Check if a URL is likely to be an external link that's incorrectly formatted as internal.
+
+    Args:
+        url (str): URL to check
+
+    Returns:
+        bool: True if the URL is likely an external link
+    """
+    parsed_url = urlparse(url)
+    path = parsed_url.path
+
+    # Common domains that might appear in the path part of malformed URLs
+    domain_patterns = [
+        "console.redhat.com",
+        "github.com",
+        "access.redhat.com",
+        "login.redhat.com",
+        ".io/",
+        ".com/",
+        ".org/",
+        ".net/",
+        "www.",
+        "http:",
+        "https:",
+    ]
+
+    # Check if the path contains any domain-like patterns
+    for pattern in domain_patterns:
+        if pattern in path:
+            return True
```

Comment:
nitpick (but feel free to ignore): `domain_patterns` can be set, not a list.

Then perhaps:

```python
return any(pattern in path for pattern in domain_patterns)
```

But dunno...